peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+
+Directly raising (a proxy of) the exception
+is problematic since it's harder to distinguish between an error
+in the ``Interpreter.exec()`` call and an uncaught exception
+from the subinterpreter.
+
+Directly Shareable Objects
+--------------------------
+
+We cannot support directly sharing any and all objects through queues
+because all objects have at least some mutable state (e.g. refcount)
+which the GIL protects normally.  Furthermore, objects must only be
+deleted by the interpreter that created them, which makes direct
+sharing even more complex.
+
+Supporting indirect sharing of all objects in queues would be
+possible by automatically pickling them if we can't use the more
+efficient mechanism.  However, it's helpful to know that only the
+efficient way is being used.  Furthermore, for mutable objects
+pickling would violate the guarantee that "shared" objects be
+equivalent (and stay that way).  Thus the proposal does not
+include automatic pickling.
+
+Objects vs. ID Proxies
+----------------------
+
+For both interpreters and queues, we provide objects that expose them
+indirectly with ID-wrapping surrogates for the underlying state.
+In both cases the state is process-global and will be used by multiple
+interpreters.  Thus they aren't suitable to be implemented as
+``PyObject``, which is only really an option for interpreter-specific
+data.  That's why the ``interpreters`` module instead provides objects
+that are weakly associated through the ID.
+
+
+Specification
+=============
+
+The module will:
+
+* expose the existing multiple interpreter support
+* introduce a basic mechanism for communicating between interpreters
+
+The module will wrap a new low-level ``_interpreters`` module
+(in the same way as the ``threading`` module).
+However, that low-level API is not intended for public use
+and thus not part of this proposal.
+
+Using Interpreters
+------------------
+
+The module defines the following functions:
+
+* ``get_current() -> Interpreter``
+      Returns an ``Interpreter`` object for the currently executing
+      interpreter.
+
+* ``list_all() -> list[Interpreter]``
+      Returns an ``Interpreter`` object for each existing interpreter,
+      whether it is currently running in any threads or not.
+
+* ``create() -> Interpreter``
+      Create a new interpreter and return an ``Interpreter`` object
+      for it.  The interpreter will not be associated with any threads
+      until something is actually run in the interpreter.
+
+Interpreter Objects
+-------------------
+
+An ``interpreters.Interpreter`` object represents the interpreter
+(``PyInterpreterState``) with the corresponding ID.
+
+If the interpreter was created with ``interpreters.create()`` then
+it will be destroyed as soon as all ``Interpreter`` objects have been
+deleted.
+
+Attributes and methods:
+
+* ``id``
+      (read-only) The target interpreter's ``InterpreterID`` object.
+      It is an int-like object with some internal bookkeeping.
```

Comment:
Why does it have to be a special object? What does the extra bookkeeping do?

Reply:
Yeah, this doesn't need to be a special object.  I'll change it to `int`.